nexus_motion_planner/src/motion_plan_cache.cpp

```diff
@@ -0,0 +1,935 @@
+        "start_state.joint_state.position_" + std::to_string(i),
+        NEXUS_MATCH_RANGE(
+          current_state_msg.joint_state.position.at(i), match_tolerance)
+      );
+    }
+  }
+  else
+  {
+    for (
+      size_t i = 0; i < plan_request.start_state.joint_state.name.size(); i++
+    )
+    {
+      query.append(
+        "start_state.joint_state.name_" + std::to_string(i),
+        plan_request.start_state.joint_state.name.at(i));
+      query.appendRangeInclusive(
+        "start_state.joint_state.position_" + std::to_string(i),
+        NEXUS_MATCH_RANGE(
+          plan_request.start_state.joint_state.position.at(i), match_tolerance)
+      );
+    }
+  }
+  return true;
+}
+
+bool
+MotionPlanCache::extract_and_append_plan_goal_to_query(
+  Query& query,
+  const moveit::planning_interface::MoveGroupInterface& /* move_group */,
+  const moveit_msgs::msg::MotionPlanRequest& plan_request,
+  double match_tolerance)
+{
+  match_tolerance += exact_match_precision_;
+
+  // Make ignored members explicit
+  bool emit_position_constraint_warning = false;
+  for (auto& constraint : plan_request.goal_constraints)
+  {
+    for (auto& position_constraint : constraint.position_constraints)
+    {
+      if (!position_constraint.constraint_region.primitives.empty())
+      {
+        emit_position_constraint_warning = true;
+        break;
+      }
+    }
+    if (emit_position_constraint_warning)
+    {
+      break;
+    }
+  }
+  if (emit_position_constraint_warning)
+  {
+    RCLCPP_WARN(
+      node_->get_logger(),
+      "Ignoring goal_constraints.position_constraints.constraint_region: "
+      "Not supported.");
+  }
+
+  // auto original = *query;  // Copy not supported.
+
+  query.appendRangeInclusive(
+    "max_velocity_scaling_factor",
+    NEXUS_MATCH_RANGE(
+      plan_request.max_velocity_scaling_factor, match_tolerance)
+  );
+  query.appendRangeInclusive(
+    "max_acceleration_scaling_factor",
+    NEXUS_MATCH_RANGE(
+      plan_request.max_acceleration_scaling_factor, match_tolerance)
+  );
+  query.appendRangeInclusive(
+    "max_cartesian_speed",
+    NEXUS_MATCH_RANGE(plan_request.max_cartesian_speed, match_tolerance));
+
+  // Extract constraints (so we don't have cardinality on goal_constraint idx.)
+  std::vector<moveit_msgs::msg::JointConstraint> joint_constraints;
+  std::vector<moveit_msgs::msg::PositionConstraint> position_constraints;
+  std::vector<moveit_msgs::msg::OrientationConstraint> orientation_constraints;
+  for (auto& constraint : plan_request.goal_constraints)
+  {
+    for (auto& joint_constraint : constraint.joint_constraints)
+    {
+      joint_constraints.push_back(joint_constraint);
+    }
+    for (auto& position_constraint : constraint.position_constraints)
+    {
+      position_constraints.push_back(position_constraint);
+    }
+    for (auto& orientation_constraint : constraint.orientation_constraints)
+    {
+      orientation_constraints.push_back(orientation_constraint);
+    }
+  }
```

Comment:
These vectors doesn't seem to be modified after these insertions, can't we avoid the copy and using `plan_request.goal_constraints` directly? (will require double loop though).

Reply:
warehouse_ros doesn't support array/vector columns.

What is happening here is we are inserting the constraints as separate columns.

E.g.:
- joint_constraint_1
- joint_constraint_2
- joint_constraint_3
- position_constraint_1
- position_constraint_2
- position_constraint_3
- orientation_constraint_1
- orientation_constraint_2
- orientation_constraint_3

If we used `goal_constraints` directly, like so, we would have:
- constraint_1
- constraint_2
- constraint_3
- constraint_4
- constraint_5
- constraint_6

The problem with doing this is it makes it much more likely that constraint ordering will matter, when we don't actually care about it when pulling up our plans.

E.g. for goal constraints that are ordered differently but still have the same constraint:
- [joint_1, position_1, joint_2, position_2]
- [joint_1, joint_2, position_1, position_2]

Then the inserted columns will have different ordering if we didn't extract out the constraint types, which means lookups will be unnecessarily sensitive to ordering (between the different types of constraints.)

Reply:
iiuc, we should also sort the constraints? e.g.
```
goal_constraints[0] = some joint_2 constraint
goal_constraints[1] = some joint_1 constraint
```
should be stored with stable ordering (`joint_1`, `joint_2`).

Reply:
Fair point!
f166c890ea598f3bbba6a1db4f0ed957352a589c